src/nvim/ex_cmds.c

```diff
@@ -801,21 +793,24 @@ int do_move(linenr_T line1, linenr_T line2, linenr_T dest)
    * their final destination at the new text position -- webb
    */
   last_line = curbuf->b_ml.ml_line_count;
-  mark_adjust(line1, line2, last_line - line2, 0L);
-  changed_lines(last_line - num_lines + 1, 0, last_line + 1, num_lines);
+  mark_adjust_nofold(line1, line2, last_line - line2, 0L);
   if (dest >= line2) {
-    mark_adjust(line2 + 1, dest, -num_lines, 0L);
+    mark_adjust_nofold(line2 + 1, dest, -num_lines, 0L);
+    foldSwapRange(&curwin->w_folds,
+        line1, line2,
+        dest + 1, dest + num_lines);
     curbuf->b_op_start.lnum = dest - num_lines + 1;
     curbuf->b_op_end.lnum = dest;
   } else {
-    mark_adjust(dest + 1, line1 - 1, num_lines, 0L);
+    mark_adjust_nofold(dest + 1, line1 - 1, num_lines, 0L);
+    foldSwapRange(&curwin->w_folds, dest + 1, line1 - 1,
+        line1, line2);
     curbuf->b_op_start.lnum = dest + 1;
     curbuf->b_op_end.lnum = dest + num_lines;
   }
   curbuf->b_op_start.col = curbuf->b_op_end.col = 0;
-  mark_adjust(last_line - num_lines + 1, last_line,
+  mark_adjust_nofold(last_line - num_lines + 1, last_line,
       -(last_line - dest - extra), 0L);
-  changed_lines(last_line - num_lines + 1, 0, last_line + 1, -extra);
```

Comment:
`changed_lines()` is needed after `mark_adjust_nofold()` -- it's called at the end of the `do_move()` function.

This is similar to what happened originally (before the previous fix was done).
Before, `mark_adjust()` was used to move marks around bit-by-bit, to get them in the correct place (because it's not a simple move everything down/up), then the extra lines from the initial copy were deleted, and `changed_lines()` was called.
Now I've replaced `mark_adjust()` with `mark_adjust_nofolds()`, and dealt with the folds independently.

This call to `changed_lines()` is a later addition that fixed part of the problem [see here](https://groups.google.com/forum/#!msg/vim_dev/5SD5E_hyGPk/-ojYGyeNbwYJ).

The single call to `foldSwapRange()` puts the folds in the correct place (assuming no bugs ...) where it is in this paragraph doesn't actually matter, as nothing else in the block modifies or uses the fold positions. I put it in the existing `if/else` because I needed an `if/else` condition on the boolean that was already there.
It may not be clear in the code, but the gist of what I've done is to completely separate the updating of the marks and folds.
If you imagine three layers -- the text, the marks, and the folds -- then `mark_adjust()` moves both the marks and folds down/up at once, while `mark_adjust_nofold()` just moves marks and `foldSwapRange()` just swaps folds.
As long as each layer lines up correctly in the end the order they're called doesn't matter, and you can call them as many times as you want.